lsat/store.go

```diff
@@ -50,42 +69,150 @@ func NewFileStore(storeDir string) (*FileStore, error) {
-func (f *FileStore) Token() (*Token, error) {
-	if !f.HasToken() {
+// currentToken returns the current token without locking the store.
+func (f *FileStore) currentToken() (*Token, error) {
+	switch {
+	case fileExists(f.fileName):
+		return readTokenFile(f.fileName)
+
+	case fileExists(f.fileNamePending):
+		return readTokenFile(f.fileNamePending)
+
+	default:
 		return nil, ErrNoToken
 	}
-	bytes, err := ioutil.ReadFile(f.fileName)
+}
+
+// AllTokens returns all tokens that the store has knowledge of, even if they
+// might be expired. The tokens are mapped by their identifying attribute like
+// file name or storage key.
+//
+// NOTE: This is part of the Store interface.
+func (f *FileStore) AllTokens() (map[string]*Token, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	tokens := make(map[string]*Token)
+
+	// All tokens start with the same name but old tokens have a timestamp
+	// appended to the original file name and the pending token has a fixed
+	// suffix.
+	tokenDir := filepath.Dir(f.fileName)
+	files, err := ioutil.ReadDir(tokenDir)
 	if err != nil {
 		return nil, err
 	}
-	return deserializeToken(bytes)
+	for _, file := range files {
+		name := file.Name()
+		if !strings.HasPrefix(name, storeFileName) {
+			continue
+		}
+		fileName := filepath.Join(tokenDir, name)
+		token, err := readTokenFile(fileName)
+		if err != nil {
+			return nil, err
+		}
+		tokens[fileName] = token
+	}
+
+	return tokens, nil
 }
 
 // StoreToken saves a token to the store, overwriting any old token if there is
 // one.
 //
 // NOTE: This is part of the Store interface.
-func (f *FileStore) StoreToken(token *Token) error {
-	bytes, err := serializeToken(token)
+func (f *FileStore) StoreToken(newToken *Token) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	// Serialize the token first, before we rename anything.
+	bytes, err := serializeToken(newToken)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(f.fileName, bytes, 0600)
+
+	// We'll need to know if there is any other token already in place,
+	// either pending or not, that we need to delete or overwrite.
+	currentToken, err := f.currentToken()
+
+	switch {
+	// No token in the store yet, just write it to the corresponding file.
+	case err == ErrNoToken:
+		// What's the target file name we are going to write?
+		newFileName := f.fileName
+		if newToken.isPending() {
+			newFileName = f.fileNamePending
+		}
+		return ioutil.WriteFile(newFileName, bytes, 0600)
+
+	// Fail on any other error.
+	case err != nil:
+		return err
+
+	// Replace a pending token with a paid one.
+	case currentToken.isPending() && !newToken.isPending():
+		// Write the new token first, so we still have the pending
+		// around if something goes wrong.
+		err := ioutil.WriteFile(f.fileName, bytes, 0600)
+		if err != nil {
+			return err
+		}
+		return os.Remove(f.fileNamePending)
+
+	// Replace a paid token with a new one by creating a backup first.
+	case !currentToken.isPending() && !newToken.isPending():
```

Comment:
This case seems unused so far since we always write the pending token first, and such an action is not allowed if we already have a prior valid token. If that's the approach we're going to take for now, I'd rather just remove this.